_drafts/unit-testing-basics.md

```diff
@@ -0,0 +1,1257 @@
+easily unit tested after becoming acquainted with best practices and writing a
+few tests of your own.
+
+<!-- excerpt start -->
+
+In this post, we go into detail on how to properly build abstractions to stub,
+fake, and mock out implementations of low level embedded software and provide a
+full real-world example of a unit test using the CppUTest 3.8 unit test
+framework.
+
+<!-- excerpt end -->
+
+This is the second post in our [Building Better Firmware
+series]({{ '/tag/better-firmware' |
+relative_url }}), following the [post]({% post_url 2019-09-17-continuous-integration-for-firmware %})
+about Continuous Integration for firmware projects, which is a wonderful pre-cursor
+to this post.
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Unit Testing Overview
+
+Unit testing is a method of testing software where individual software
+components are isolated and tested for correctness. Ideally, these unit tests
+are able to cover most if not all of the code paths, argument bounds, and
+failure cases of the software under test.
+
+Through proper use of unit tests, and especially while using practices from Test
+Driven Development (TTD)[^0], the time it takes to **stabilize** embedded
+software can decrease dramatically, making individuals and teams more productive
+and firmware less likely to experience functional bugs, control flow bugs, and
+even fatal issues, such as memory leaks and (gasp!) bootloops.
+
+{:.no_toc}
+
+### Life Before Unit Testing
+
+Here are a few examples that I've experienced in the past that were alleviated
+by the team doubling down on unit testing the firmware.
+
+- You find testing on target hardware **slow** and **inconvenient**, especially
+  when multiple devices (e.g. a mobile phone) or prior setup (e.g. a factory
+  reset) is required
+- Bugs and regressions occur **repeatedly** in a single piece of software.
+- Deadlocks, HardFaults and Memory Leaks are the norm and have become accepted
+  (unit testing example for preventing deadlocks included below!).
+- The amount of time spent debugging and testing firmware is **60% or
+  more**[^1].
+- The first instinct when starting a new software module is to write a chunk of
+  code and **test on hardware**.
+
+{:.no_toc}
+
+### Life After Unit Testing (Possibly)
+
+At a previous company, after scrapping most legacy code and writing new modules
+with 90%+ code coverage and through the use of TTD, this is what development
+felt like sometimes.
+
+- You write a new file, maybe an in-flash log storage module, and it works the
+  **first time** when flashed on the device (no better feeling).
+- Regressions are caught **immediately** when runnings tests locally or in CI.
+- Memory leaks are raised as errors in unit tests.
+- Testing a majority of the firmware only takes **a minute**.
+- Writing and debugging code on a plane is possible! No hardware required.
```

Comment:
One strong upside for TDD is that IMO, it often results in a cleaner, better structured code because you are forced to put related behavior in module with clear boundaries vs than having everything rely on everything.

Reply:
Added!